rfd/XXXX-upgrade-system.md

```diff
@@ -0,0 +1,659 @@
+
+- Installation targets are arbitrary mappings that must *at least* contain `version`, but may contain
+any additional information as well. Certain metadata is understood by teleport (e.g. `fips:yes|no`,
+`arch:amd64|arm64|...`), but additional metadata (e.g. `sha256sum:12345...`) is simply passed through
+to the install controller.  The target to be used for a given server is the first target that
+*is not incompatible* (i.e. no attempt to find the "most compatible" target is made).  A target is
+incompatible with a server if that server's version cannot safely upgrade/downgrade to that target,
+*or* if the target specifies a build attribute that differs from a build attribute of the current
+installation (e.g. `fips:yes` when current build is `fips:no`). We don't require that all build attributes
+are present since not all systems require knowledge of said attributes.  It is the responsibility of
+an installer to fail if it is not provided with sufficient attributes to perform the installation
+safely.
+
+
+- Reconciliation loop high-level steps:
+  - Load `version-directive` resource.
+  - Match connected servers to directives via selector.
+  - Match targets from appropriate directives to servers based on currently installed target.
+  - Filter out servers which already have the correct target installed.
+  - Pass remaining servers to first compatible install controller.
+
+TODO
+
+```yaml
+# version directive is a singleton resource that is either supplied by a user,
+# or periodically generated by a version controller (e.g. tuf, plugin, etc).
+# this represents the desired state of the cluster, and is used to guide a control
+# loop that matches install targets to appropriate nodes and installers.
+kind: version-directive
+version: v1
+metadata:
+  name: version-directive
+spec:
+  nonce: 2
+  status: enabled
+  version_controller: static-config
+  stale_after: <time>
+  not_before: <time>
+  not_after: <time>
+  directives:
+    - name: Staging
+      targets:
+        - version: 2.3.4
+          fips: yes
+        - version: 2.3.4
+          fips: no
+      installers:
+        - kind: script
+          name: apt-install
+      selectors:
+        - labels:
+            env: staging
+        - labels:
+            env: testing
+
+    - name: Prod
+      targets:
+        - version: 1.2.3
+          fips: yes
+      installers:
+        - kind: script
+          name: apt-install
+      selectors:
+        - labels:
+            env: prod
+
+    - name: Minimum
+      targets:
+        - version: 1.2.0
+          fips: no
+      installers:
+        kind: script
+        name: apt-install
+      selectors:
+        - labels:
+            '*': '*'
+```
+
+
+### Version Controllers
+
+- A `version-controller` is a loop that periodically generates a `version-directive`.  It may
+be a loop that runs within the auth server, or an external plugin.
+
+- The only currently planned `version-controller` is the TUF version controller.
```

Comment:
@reedloden would be cool for us to think through possible attacks on this model of upgrade. Can we mount an attack to keep Teleport on stale versions for example.